src/crimson/net/SocketConnection.cc

```diff
@@ -312,6 +313,23 @@ seastar::future<> SocketConnection::send(MessageRef msg)
   return f.get_future();
 }
 
+seastar::future<> SocketConnection::keepalive()
+{
+  seastar::shared_future<> f = send_ready.then([this] {
+      const char tag = CEPH_MSGR_TAG_KEEPALIVE2;
+      const auto stamp = ceph::coarse_real_clock::to_ceph_timespec(
+        ceph::coarse_real_clock::now());
+      auto msg = seastar::net::packet{&tag, sizeof(tag)};
+      msg.append(seastar::net::packet{reinterpret_cast<const char*>(&stamp),
+	                              sizeof(stamp)});
```

Comment:
these packets have pointers to stack memory that goes out of scope (here and in `handle_keepalive2()`)

for the handshake logic, i worked around this issue by adding a `struct Handshake` to `SocketConnection` so there's always a single instance available - this works as long as you never have multiple outgoing handshake messages. i see there's already a `struct Keepalive` that should help with this part?

i'm also curious how much we can cheat here - if we know that there's an actual message pending, is there any point in sending the keepalive message at all? what do you think about this shortcut?
```c++
  if (!send_ready.available()) {
    // still connecting or a message is already queued
    return seastar::make_ready_future();
  }
```

Reply:
@cbodley yeah. i thought the `packet()` constructor was different than `packet::from_static_data()`. but i was wrong.

i am using a packed structure for the keepalive messages in the latest change.

no, there is not =) but i checked the receiver side's handling of  `CEPH_MSGR_TAG_KEEPALIVE2` messages in hope that it can update the keep alive timestamp when it receives any messages. but it turns out it does not take non-keepalive messages into account when updating `last_keepalive_ack` or `last_keepalive`.

so unless we update the receiver side, i am afraid that we cannot cheat this way.